torch/csrc/autograd/type_and_shape.h

```diff
@@ -3,16 +3,37 @@
 #include <ATen/ATen.h>
 #include "torch/csrc/assertions.h"
 
+#ifdef WITH_CUDA
+  #include "THC/THCStream.h"
+#endif 
+
 namespace torch { namespace autograd {
 
 /// A tensor's type and shape. Each Function records the required type and
 /// shape of its inputs. If is_valid() is false, then the corresponding input
 /// is not used and may be an undefined tensor.
 struct TypeAndShape {
-  TypeAndShape() : type_(nullptr) {}
+  static constexpr int NO_DEVICE = -2;
+
+  TypeAndShape() : type_{nullptr}, device_{NO_DEVICE} {}
```

Comment:
This is no longer just `TypeAndShape`

Reply:
Also, can we make it constructible from a tensor? This would let us simplify the `add_input_metadata` call to take an `at::Tensor`, and we could avoid changing all the call sites in case of future changes like this one.

Reply:
I was hoping you'd suggest a name ;)

How about "InputMetadata?" I really like the idea of making it constructible from a Tensor, let me see about doing that.